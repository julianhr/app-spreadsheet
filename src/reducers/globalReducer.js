import { createReducer } from 'redux-starter-kit'

export const INITIAL_STATE = {
  activeCell: null,
  rows: 14,
  columns: 6,
}

export default createReducer(INITIAL_STATE, {
  'SET_ACTIVE_CELL': (state, { payload }) => {
    state.activeCell = payload
  },
})
