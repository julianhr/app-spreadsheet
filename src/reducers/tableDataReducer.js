import { createReducer } from 'redux-starter-kit'


const INITIAL_STATE = {}

export default createReducer(INITIAL_STATE, {
  'SET_CELL_DATA': (state, { payload }) => {
    const { location, data } = payload
    state[location] = data
  },

  'DELETE_CELL_DATA': (state, { payload }) => {
    delete state[payload]
  }
})
