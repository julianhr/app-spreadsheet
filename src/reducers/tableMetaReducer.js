import { createReducer } from 'redux-starter-kit'
import { DEFAULT_COL_WIDTH } from '~/library/constants'


const INITIAL_STATE = {
  colWidths: {},
}

export default createReducer(INITIAL_STATE, {
  'SET_COL_WIDTH': (state, { payload }) => {
    const { colLabel, width } = payload
    const { colWidths } = state
    
    if (width === DEFAULT_COL_WIDTH) {
      delete state.colWidths[colLabel]
    } else {
      state.colWidths = { ...colWidths, [colLabel]: width }
    }
  },
})
