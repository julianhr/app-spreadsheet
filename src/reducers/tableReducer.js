import { createReducer } from 'redux-starter-kit'


const INITIAL_STATE = {}

export default createReducer(INITIAL_STATE, {
  'SET_CELL_VALUE': (state, { payload }) => {
    const { location, text, formula } = payload
    state[location] = { text, formula }
  },

  'DELETE_CELL_VALUE': (state, { payload }) => {
    delete state[payload]
  }
})
