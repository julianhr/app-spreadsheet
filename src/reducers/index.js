import { configureStore } from 'redux-starter-kit'

import globalReducer from './globalReducer'
import tableReducer from './tableReducer'


const appStoreFn = () => (
  configureStore({
    reducer: {
      global: globalReducer,
      table: tableReducer,
    }
  })
)


export default appStoreFn
