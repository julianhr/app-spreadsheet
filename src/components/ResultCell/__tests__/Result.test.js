import React from 'react'
import { render, fireEvent, cleanup } from '@testing-library/react'

import { appStoreGen } from '~/reducers/'
import MockApp from '~/__tests__/__mocks__/MockApp'
import ReduxResult, { Result } from '../Result'


const requiredProps = {
  location: 'A-3',
  fwdRef: React.createRef(),
  // redux
  openFloatingInputter: jest.fn(),
  clearCellData: jest.fn(),
  setActiveCell: jest.fn(),
  result: 5,
  isActive: true,
}

const testProps = {
  ...requiredProps,
}

const renderApp = (props, customStore) => {
  const store = customStore || appStoreGen()
  const wrapper = render(
    <MockApp customStore={store}>
      <Result {...props} />
    </MockApp>
  )

  return [wrapper, store]
}

describe('Result', () => {
  afterEach(cleanup)

  const selector = '[data-cell="result"]'

  test('required props', () => {
    Object.keys(requiredProps).forEach(prop => {
      const props = { ...requiredProps }
      delete props[prop]
      expect(() => renderApp(props)).toThrow()
    })
  })

  describe('#openFloatingInputter', () => {
    test('matches signature', () => {
      renderApp(testProps)
      const el = document.querySelector(selector)
      jest.spyOn(testProps, 'openFloatingInputter')
      fireEvent.doubleClick(el)
      expect(testProps.openFloatingInputter.mock.calls).toMatchSnapshot()
    })
  })

  describe('#handleOnFocus', () => {
    test('calls #setActiveCell', () => {
      renderApp(testProps)
      const el = document.querySelector(selector)
      jest.spyOn(testProps, 'setActiveCell')
      fireEvent.focus(el)
      expect(testProps.setActiveCell).toHaveBeenCalledWith(testProps.location)
    })
  })

  describe('#handleOnClick', () => {
    test('stops event propagation', () => {
      const spyOnClick = jest.fn()
      const ListenClick = ({ children }) => (<div onClick={spyOnClick}>{children}</div>)
      render(
        <MockApp>
          <ListenClick>
            <Result {...testProps} />
          </ListenClick>
        </MockApp>
      )
      const el = document.querySelector(selector)
      jest.spyOn(testProps, 'setActiveCell')
      fireEvent.click(el)
      expect(spyOnClick).not.toHaveBeenCalled()
    })
  })

  describe('#handleOnDoubleClick', () => {
    test('calls #openFloatingInputter', () => {
      renderApp(testProps)
      const el = document.querySelector(selector)
      jest.spyOn(Result.prototype, 'openFloatingInputter')
      fireEvent.doubleClick(el)
      expect(Result.prototype.openFloatingInputter).toHaveBeenCalledWith(false)
    })

    test('stops event propagation', () => {
      const spyOnDoubleClick = jest.fn()
      const ListenClick = ({ children }) => (
        <div onCDoublelick={spyOnDoubleClick}>{children}</div>
      )
      render(
        <MockApp>
          <ListenClick>
            <Result {...testProps} />
          </ListenClick>
        </MockApp>
      )
      const el = document.querySelector(selector)
      jest.spyOn(testProps, 'setActiveCell')
      fireEvent.doubleClick(el)
      expect(spyOnDoubleClick).not.toHaveBeenCalled()
    })
  })

  describe('#handleOnKeyDown', () => {
    it('calls #clearCellData if key pressed is Backspace or Delete', () => {
      renderApp(testProps)
      const el = document.querySelector(selector)
      jest.spyOn(testProps, 'clearCellData')

      fireEvent.keyDown(el, { key: 'Backspace' })
      expect(testProps.clearCellData).toHaveBeenCalledTimes(1)
      expect(testProps.clearCellData).toHaveBeenCalledWith(testProps.location)
      
      fireEvent.keyDown(el, { key: 'Delete' })
      expect(testProps.clearCellData).toHaveBeenCalledTimes(2)
      expect(testProps.clearCellData).toHaveBeenCalledWith(testProps.location)
    })

    it('calls #openFloatingInputter if key pressed is a character', () => {
      renderApp(testProps)
      const el = document.querySelector(selector)
      jest.spyOn(Result.prototype, 'openFloatingInputter')

      fireEvent.keyDown(el, { key: 'a' })
      expect(Result.prototype.openFloatingInputter).toHaveBeenCalledWith(true)
    })
  })

  describe('functional tests', () => {
    it('matches snapshot with Redux', () => {
      const wrapper = render(
        <MockApp>
          <ReduxResult {...testProps} />
        </MockApp>
      )
      expect(wrapper.asFragment()).toMatchSnapshot()
    })
  })
})
