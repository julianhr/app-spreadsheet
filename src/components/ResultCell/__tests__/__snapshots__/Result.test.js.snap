// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result #openFloatingInputter matches signature 1`] = `
Array [
  Array [
    Object {
      "bottom": 0,
      "height": 0,
      "left": 0,
      "right": 0,
      "top": 0,
      "width": 0,
    },
    false,
  ],
]
`;

exports[`Result functional tests matches snapshot with Redux 1`] = `
<DocumentFragment>
  <div
    class="css-1516bbv-Root-Datum e1rvd88v0"
    data-cell="result"
    data-location="A-3"
    tabindex="0"
  />
</DocumentFragment>
`;
